### PR TITLE
fix: * will match when value is not empty

### PR DESCRIPTION
**Pull request by yasinlachiny**
**Please provide a description of this PR:**
According to [Istio document](https://istio.io/latest/docs/reference/config/security/authorization-policy/)  `“*”` will match when value is not empty. But right now it's not working correctly because it's using  `*` but it should use `.+` like the [StringMatcher](https://github.com/istio/istio/blob/d0705cf0ed5591cc26c08001f3faab0a880aec48/pilot/pkg/security/authz/matcher/string.go#L50)  

For example, we want to block an empty header but allow a nonempty header.
We can create an `AuthorizationPolicy`
```
kind: AuthorizationPolicy
metadata:
  name: test
  namespace: istio-system
spec:
  action: DENY
  rules:
  - from:
    when:
    - key: request.headers[test]
      notValues: ["*"]
```
and then try to access it.
`curl -v -H 'test;' DOMAIM
`
`curl -v -H 'test: "nonempty"' DOMAIM`

By using this PR it will block the empty header and allow nonempty header. `“*” `will match when value is not empty

### CODE REVIEW

File: pilot/pkg/security/authz/matcher/header.go [hunk 1]
```diff
@@ -23,14 +23,20 @@ import (
...
 // HeaderMatcher converts a key, value string pair to a corresponding HeaderMatcher.
+// The wildcard "*" will be generated as ".+" instead of ".*".
 func HeaderMatcher(k, v string) *routepb.HeaderMatcher {
 	// We must check "*" first to make sure we'll generate a non empty value in the prefix/suffix case.
 	// Empty prefix/suffix value is invalid in HeaderMatcher.
 	if v == "*" {
 		return &routepb.HeaderMatcher{
 			Name: k,
-			HeaderMatchSpecifier: &routepb.HeaderMatcher_PresentMatch{
-				PresentMatch: true,
```
Comment (hzxuzhonghu) [inline pilot/pkg/security/authz/matcher/header.go:33]: If this is removed, how could we express if a header exists?

Comment (yasinlachiny) [reply pilot/pkg/security/authz/matcher/header.go:33]: Hi @hzxuzhonghu 
Thank you for reviewing my code.

Based on [Istio document](https://istio.io/latest/docs/reference/config/security/authorization-policy/)

> Presence match: “*” will match when value is not empty.

Present match means the value is not empty. Right now it checks if a header exists that's against the document and also the rest of code like [StringMatcher](https://github.com/istio/istio/blob/d0705cf0ed5591cc26c08001f3faab0a880aec48/pilot/pkg/security/authz/matcher/string.go#L50)

Comment (hzxuzhonghu) [reply pilot/pkg/security/authz/matcher/header.go:33]: Yes, i understand the context. This is the question most user will face, we need to consider carefully

**Review by howardjohn — CHANGES_REQUESTED** [review]
Putting a hold since this is changing behavior of authz policies which may have massive implications for users upon upgrade.

cc @aryan16 @liminw @costinm

**Comment by yasinlachiny** [comment]
Hi @howardjohn 

Could you please update me about this change?

Right now I implemented this functionality with Envoyfilter and that's too messy. If we could have this change in the next version that would be great.

**Comment by howardjohn** [comment]
Changing an API like this may appear to be a simple bug fix, but since we are changing the behavior of a stable API - especially such a security sensitive one - its extremely tricky and sometimes best to not change it.

I am not the owner for this part of the code though, @liminw and @aryan16 should comment.

**Comment by yasinlachiny** [comment]
Hi @liminw @aryan16 
Could you please share your opinion about this bug fix?

**Comment by yasinlachiny** [comment]
lifecycle/needs-escalatio

**Comment by yasinlachiny** [comment]
lifecycle/needs-escalation

**Comment by aryan16** [comment]
I agree with John on this, there might be some cases where users are intentionally using this as a feature (just checking whether header is present in the request and once it is there, they would add some value to it and pass it in some other request/response). It shouldn't be feasible to just update the behaivor like this, could be a breaking change. If we would want to support this case, should we add one more extra specifier (eg. `.+`) without changing the current specifier (`*`)? That would resolve this issue for new users and won't break the existing ones? cc : @howardjohn, @liminw

**Pull request closed without merging**